chain/network/src/peer_manager/network_state/routing.rs

```diff
@@ -164,4 +184,40 @@ impl NetworkState {
     pub(crate) fn compare_route_back(&self, hash: CryptoHash, peer_id: &PeerId) -> bool {
         self.tier2_route_back.lock().get(&hash).map_or(false, |value| value == peer_id)
     }
+
+    /// Accepts NetworkTopologyChange events.
+    /// Changes are batched via the `update_routes_demux`, then passed to the V2 routing table.
+    /// If an updated DistanceVector is returned by the routing table, broadcasts it to peers.
+    /// If an error occurs while processing a DistanceVector advertised by a peer, bans the peer.
+    pub async fn update_routes(
+        self: &Arc<Self>,
+        clock: &time::Clock,
+        event: NetworkTopologyChange,
+    ) -> Result<(), ReasonForBan> {
+        let this = self.clone();
+        let clock = clock.clone();
+        self.update_routes_demux
+            .call(event, |events: Vec<NetworkTopologyChange>| async move {
+                let (to_broadcast, oks) =
+                    this.graph_v2.batch_process_network_changes(&clock, events).await;
+
+                match to_broadcast {
+                    Some(my_distance_vector) => {
+                        // TODO: should we put some event type here?
+                        // this.config.event_sink.push(Event::EdgesAdded(spt.edges.clone()));
+                        this.broadcast_distance_vector(my_distance_vector);
+                    }
+                    None => {}
+                }
+
+                oks.iter()
+                    .map(|ok| match ok {
+                        true => Ok(()),
+                        false => Err(ReasonForBan::InvalidDistanceVector),
+                    })
+                    .collect()
+            })
+            .await
+            .unwrap_or(Ok(()))
```

Comment:
I don't quite get what's going on here. It looks like oks are mapped to some Ok-s or Err-s and then returned? How does that vector/slice get converted to the return type of this function which is ReasonForBan? 

Also you unwrap_or(Ok) - so if an error happend you return a Ok?

Reply:
This function gets called with a single event to be processed. The `update_routes_demux.call` takes a closure describing how to process a batch of events. Inside the closure, a vector of events is mapped to a vector of `Result<(), ReasonForBan>`. Outside the demux call, the single `Result<(), ReasonForBan>` corresponding to the single event passed in from this call is returned. 

The result from the demux is wrapped inside a `Result<Res, ServiceStoppedError>>` ([see link](https://github.com/near/nearcore/blob/a262c2a709a6a1864ca7de91a1d3efbe91a9e88d/chain/network/src/concurrency/demux.rs#L92C44-L92C77)). If we do get back a `ServiceStoppedError`, it is not a reason to ban the peer, hence the `unwrap_or(Ok())`.